_posts/jvm-debug-symbols.md

```diff
@@ -0,0 +1,285 @@
+
+And this is calling the JVM internally to [`java_lang_Throwable::fill_in_stack_trace`](https://github.com/openjdk/jdk/blob/5d5bf16b0af419781fd336fe33d8eab5adf8be5a/src/hotspot/share/classfile/javaClasses.cpp#L2403-L2538)
+
+and store it into the [`backtrace`](https://github.com/openjdk/jdk/blob/12dca36c73583d0ed2e1f684b056100dc1f2ef55/src/java.base/share/classes/java/lang/Throwable.java#L122) field in `Throwable` class a list of pointer (or handle) to method metadata from the interpreter state and the BCI (ByteCode Index).
+
+Then, either if you call `getStackTrace()` or `printStackTrace()` on an exception, it will take this backtrace to fill out StackTraceElement array:
+[`java_lang_Throwable::get_stack_trace_elements`](https://github.com/openjdk/jdk/blob/1581e3faa06358f192799b3a89718028c7f6a24b/src/hotspot/share/classfile/javaClasses.cpp#L2608-L2643)
+and try to resolve symbol names and line numbers with the help of [`ConstantPool::source_file_name`](https://github.com/openjdk/jdk/blob/5d5bf16b0af419781fd336fe33d8eab5adf8be5a/src%2Fhotspot%2Fshare%2Foops%2FconstantPool.hpp#L195-L198) to fetch the source file name from the constant pool of the class and [`Method::line_number_from_bci`](https://github.com/openjdk/jdk/blob/5d5bf16b0af419781fd336fe33d8eab5adf8be5a/src/hotspot/share/oops/method.cpp#L940-L962)
+to use the LineNumber Table to translate BCI to line number.
+
+For the interpreter it seems obvious that there is a 1:1 mapping between the current state of execution of the bytecode and the source file/line number. but for JITed code?
+
+## C2 JIT compiler
+When compiling a method, a [debug information recorder](https://github.com/openjdk/jdk/blob/5d5bf16b0af419781fd336fe33d8eab5adf8be5a/src/hotspot/share/opto/compile.cpp#L968) is started and at each method call, a safepoint is inserted. 
+
+See the [comment](https://github.com/openjdk/jdk/blob/5d5bf16b0af419781fd336fe33d8eab5adf8be5a/src/hotspot/share/code/debugInfoRec.hpp#L40-L63) self-explains:
+```
+//** The DebugInformationRecorder collects debugging information
+//   for a compiled method.
+//   Debugging information is used for:
+//   - garbage collecting compiled frames
+//   - stack tracing across compiled frames
+//   - deoptimizating compiled frames
+```
+
+A safepoint is point in code where it is safe for application threads to be stopped for doing some VM operations like GC that needs to inspect thread stacks.
+When thread execution is stopped at those safepoints, JVM state is perfectly known: local variables & registers may contain reference to object that needs to be tracked.
+
+Those safepoints are emitted by the JIT compiler at strategic places that balance the execution speed and reactivity to stop threads. It's also a trade-off for debug information recording as we cannot keep track of all machine instructions and their mapping equivalent to BCI/source line numbers.
```

Comment:
```suggestion
Those safepoints are emitted by the JIT compiler at strategic places that balance the execution speed and reactivity to suspending the thread. It's also a trade-off for debug information recording since we cannot keep track of the mapping between all machine instructions and their equivalent BCI/source line numbers.
```